### PR TITLE
ARROW-8764: [C++] Make executor configurable in ReadAsync and ReadRangeCache

### PR DESCRIPTION
**Pull request by pitrou**

### CODE REVIEW

**Review by lidavidm — APPROVED** [review]
Looks good to me. Thanks for doing this!

File: cpp/src/arrow/io/interfaces.cc [hunk 1]
```diff
@@ -40,10 +40,15 @@
 
 namespace arrow {
 
+using internal::Executor;
 using internal::ThreadPool;
 
 namespace io {
 
+Executor* AsyncContext::GetExecutor() const {
+  return (executor == nullptr) ? internal::GetIOThreadPool() : executor;
```
Comment (fsaintjacques) [inline cpp/src/arrow/io/interfaces.cc:49]: Can you move this in the constructor of AsyncContext?

**Review by fsaintjacques — APPROVED** [review]
LGTM!

File: cpp/src/arrow/io/interfaces.h [hunk 2]
```diff
@@ -51,6 +48,15 @@ struct ReadRange {
   }
 };
 
+// EXPERIMENTAL
+struct AsyncContext {
```
Comment (fsaintjacques) [inline cpp/src/arrow/io/interfaces.h:52]: Windows failure points to a missing ARROW_EXPORT.

**Pull request closed without merging**